src/Tools/dotnet-monitor/CollectionRules/CollectionRuleService.cs

```diff
@@ -232,5 +239,120 @@ private async Task StopAndDisposeContainerAsync(CancellationToken token)
                 await container.DisposeAsync();
             }
         }
+
+        public Dictionary<string, CollectionRuleDescription> GetCollectionRulesDescriptions(IEndpointInfo endpointInfo)
+        {
+            Dictionary<string, CollectionRuleDescription> collectionRulesDescriptions = new();
+
+            if (_containersMap.TryGetValue(endpointInfo, out CollectionRuleContainer container))
+            {
+                container.Pipelines.ForEach(pipeline => collectionRulesDescriptions.Add(pipeline.Context.Name, GetCollectionRuleDescription(pipeline)));
+            }
+
+            return collectionRulesDescriptions;
+        }
+
+        public CollectionRuleDetailedDescription GetCollectionRuleDetailedDescription(string collectionRuleName, IEndpointInfo endpointInfo)
+        {
+            if (_containersMap.TryGetValue(endpointInfo, out CollectionRuleContainer container))
+            {
+                IEnumerable<CollectionRulePipeline> pipelines = container.Pipelines.Where(pipeline => pipeline.Context.Name.Equals(collectionRuleName));
+
+                if (null == pipelines || !pipelines.Any())
+                {
+                    throw new ArgumentException(string.Format(CultureInfo.InvariantCulture, Strings.ErrorMessage_CollectionRuleNotFound, collectionRuleName));
+                }
+
+                return GetCollectionRuleDetailedDescription(pipelines.First());
+            }
+
+            return null;
+        }
+
+        internal static CollectionRuleDescription GetCollectionRuleDescription(CollectionRulePipeline pipeline)
+        {
+            CollectionRuleLimitsOptions limitsOptions = pipeline.Context.Options.Limits;
+
+            DateTime currentTime = pipeline.Context.Clock.UtcNow.UtcDateTime;
+            pipeline.DequeueOldTimestamps(limitsOptions?.ActionCountSlidingWindowDuration, currentTime);
+
+            CollectionRuleStateHolder stateHolderCopy = pipeline.GetStateHolderCopy();
```

Comment:
We may have discussed this already, but we should consider locking during the copy operation. Without locking, it would be possible to get a new state but an old reason for example.